core/shared/src/main/scala/scalaz/zio/RefM.scala

```diff
@@ -66,15 +66,15 @@ object RefM extends Serializable {
     update: A => IO[Nothing, (B, A)],
     promise: Promise[Nothing, B]
   ) {
-    final def run(a: A): IO[List[Throwable], A] =
+    final def run(a: A, ref: Ref[A]): IO[List[Throwable], Unit] =
       interrupted.get.flatMap {
         case Some(ts) => IO.fail(ts)
         case None =>
           update(a).sandboxed.redeem({
             case Left(ts) => IO.fail(ts)
             case Right(n) => n
           }, {
-            case (b, a) => promise.complete(b).const(a)
+            case (b, a) => ref.set(a) <* promise.complete(b)
```

Comment:
I moved `ref.set` because there was a race condition making the tests fail occasionally.
`ref.set` must be done before completing the promise, otherwise `refM.modify` returns before the value is actually set.

Reply:
Submitted a dedicated PR #343